### PR TITLE
ui-core: timepicker handles seconds

### LINKED ISSUE #602: timepicker: handle seconds
### Description and goal

Add seconds to the timePicker

### Acceptance criteria

- the timePicker can show seconds

### PR DESCRIPTION
**Pull request by anisometropie**
closes #602

### CODE REVIEW

File: ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts
Comment (clarani) [inline ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts:1]: You can remove these changes (it has been solved in a previous PR)

**Review by clarani — CHANGES_REQUESTED** [review]
I think there is a problem on the css
image

File: ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts
Comment (anisometropie) [reply ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts:1]: I probably messed up a rebase at some point

File: ui-core/src/components/inputs/TimePicker.tsx [hunk 1]
```diff
@@ -47,23 +64,56 @@ const TimePicker = ({ onTimeChange, hours = 0, minutes = 0, ...otherProps }: Tim
...
+    let newMinutes = minutes;
+    let newHours = hours;
+
+    if (newSeconds >= 60) {
+      newSeconds = 0;
+      newMinutes += 1;
+
+      if (newMinutes >= 60) {
+        newMinutes = 0;
+        newHours = newHours === 23 ? 0 : newHours + 1;
```
Comment (SarahBellaha) [inline ui-core/src/components/inputs/TimePicker.tsx]:
```suggestion
        newHours = (newHours + 1) % 24;
```
Shorter but maybe less obvious ?

**Review by SarahBellaha — APPROVED** [review]
Just left a tiny suggestion. Works great, lgtm :)

File: ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts
Comment (anisometropie) [reply ui-speedspacechart/src/components/helpers/drawElements/speedLimits.ts:1]: done

File: ui-core/src/components/inputs/TimePicker.tsx (hunk 1 above)
Comment (anisometropie) [reply ui-core/src/components/inputs/TimePicker.tsx]: no, it’s better. done

File: ui-core/src/components/inputs/TimePicker.tsx [hunk 2]
```diff
@@ -74,12 +124,16 @@ const TimePicker = ({ onTimeChange, hours = 0, minutes = 0, ...otherProps }: Tim
   return (
     <div className="time-picker">
       <Input
+        className={cx('input', 'time-input', {
+          'time-input-firefox': navigator.userAgent.includes('Firefox'),
```
Comment (kmer2016) [inline ui-core/src/components/inputs/TimePicker.tsx]: What do you think about applying the style of `time-input-firefox` class using the css rules `@-moz-document url-prefix()`? For example :
```css
@-moz-document url-prefix() {
  .time-picker .time-input {
    letter-spacing: -0.9px;
  }
}
```
This way, you wouldn’t need to add a specific class for Firefox, and the style would only apply to Firefox without changing the component.

**Review by kmer2016 — COMMENTED** [review]
Nice ! just a small suggestion

File: ui-core/src/components/inputs/TimePicker.tsx (hunk 2 above)
Comment (anisometropie) [reply ui-core/src/components/inputs/TimePicker.tsx]: it’s much better ! done

**Review by kmer2016 — APPROVED** [review]

**Comment by anisometropie** [comment]
After discussion with @thibautsailly, issues with AM/PM is not a priority, considering virtually none of our users are from region using AM/PM

File: ui-core/src/components/inputs/TimePicker.tsx [hunk 3]
```diff
@@ -42,28 +59,61 @@ const TimePicker = ({ onTimeChange, hours = 0, minutes = 0, ...otherProps }: Tim
...
+        newMinutes = 0;
+        newHours = (newHours + 1) % 24;
+      }
+    } else if (newSeconds < 0) {
+      newSeconds = 59;
+      newMinutes -= 1;
+
+      if (newMinutes < 0) {
+        newMinutes = 59;
+        newHours = (newHours + 23) % 24; // minus 1 hour
```
Comment (clarani) [inline ui-core/src/components/inputs/TimePicker.tsx:89]:
```suggestion
        newHours = (newHours - 1) % 24;
```

File: ui-core/src/components/inputs/TimePicker.tsx [hunk 4]
```diff
@@ -42,28 +59,61 @@ const TimePicker = ({ onTimeChange, hours = 0, minutes = 0, ...otherProps }: Tim
 
     if (newMinutes >= 60) {
       newMinutes = 0;
-      newHours = newHours === 23 ? 0 : newHours + 1;
+      newHours = (newHours + 1) % 24;
     } else if (newMinutes < 0) {
       newMinutes = 59;
-      newHours = newHours === 0 ? 23 : newHours - 1;
+      newHours = (newHours + 23) % 24; // minus 1 hour
```
Comment (clarani) [inline ui-core/src/components/inputs/TimePicker.tsx:65]:
```suggestion
      newHours = (newHours - 1) % 24;
```

File: ui-core/src/components/inputs/TimePicker.tsx [hunk 5]
```diff
@@ -32,7 +42,14 @@ const TimeRange = ({ range, selectedItem, className, onSelectItem }: TimeRangePr
   </div>
 );
 
-const TimePicker = ({ onTimeChange, hours = 0, minutes = 0, ...otherProps }: TimePickerProps) => {
+const TimePicker = ({
+  onTimeChange,
+  hours = 0,
+  minutes = 0,
+  seconds = 0,
+  displaySeconds,
```
Comment (clarani) [inline ui-core/src/components/inputs/TimePicker.tsx]: Can you make displaySeconds false by default ?
```suggestion
  displaySeconds = false,
```

File: ui-core/src/stories/TimePicker.stories.tsx [hunk 6]
```diff
@@ -70,11 +68,39 @@ export const Default: Story = {
   args: {
     label: 'Time',
   },
+  decorators: [
+    (Story) => (
+      <div style={{ maxWidth: '6.7rem', minHeight: '500px' }}>
```
Comment (clarani) [inline ui-core/src/stories/TimePicker.stories.tsx:73]: Why not add this div in the story directly ?

File: ui-core/src/styles/inputs/timePicker.css [hunk 7]
```diff
@@ -1,9 +1,11 @@
-.time-picker {
-  .time-input {
-    cursor: pointer;
+@-moz-document url-prefix() {
```
Comment (clarani) [inline ui-core/src/styles/inputs/timePicker.css:1]: What is this `@-moz-document url-prefix` ? Is it a class specific for mozilla ?
Anyway, can you move this code in .time-picker l7 ?

File: ui-core/src/components/inputs/TimePicker.tsx (hunk 3 above)
Comment (anisometropie) [reply ui-core/src/components/inputs/TimePicker.tsx:89]: `-1 % 24` is just `-1`. it doesn’t roll back to 23

File: ui-core/src/components/inputs/TimePicker.tsx (hunk 4 above)
Comment (anisometropie) [reply ui-core/src/components/inputs/TimePicker.tsx:65]: idem, can’t do that

File: ui-core/src/components/inputs/TimePicker.tsx (hunk 5 above)
Comment (anisometropie) [reply ui-core/src/components/inputs/TimePicker.tsx]: yes, done

File: ui-core/src/stories/TimePicker.stories.tsx (hunk 6 above)
Comment (anisometropie) [reply ui-core/src/stories/TimePicker.stories.tsx:73]: I think it’s the purpose of the decorator. that’s what’s done in the rest of the stories. I don’t think it’s worth dealing with more props

File: ui-core/src/styles/inputs/timePicker.css (hunk 7 above)
Comment (anisometropie) [reply ui-core/src/styles/inputs/timePicker.css:1]: yes, the selector only applies to firefox.
I tried to move this in the .time-picker block but it just breaks the style.

I think it doesn’t work if it’s not at the top level

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:

**Pull request merged**